### DISCUSSION TITLE
How can I generate a Business Model Canvas (BMC) with text from a dataframe within a qmd-document?

### DISCUSSION THREAD

**Discussion by sebastiansukstorf** [quarto-dev/quarto-cli#9051] [Q&A] [answered]
### Description

Hello,
I need to generate a Business Model Canvas (BMC) image or graph with text from a dataframe within a qmd-document. I tried to use mermaid or D2 diagram. Another way I tried was to generate a markdown table.  The output from mermaid and D2 is not fine for my. The markdown table is also not fine. Are there any other possiblities to generate a graph or diagram with data? Any help ist wellcome. Sebastian
![business-model-canvas-eng-big](https://github.com/quarto-dev/quarto-cli/assets/6294081/2c071ea1-b447-4820-9249-d3c4c84ee520)

**Comment by mcanouil** [comment]
Did you try the project type `dashboard`? See <https://quarto.org/docs/dashboards/>.

As for "with text from a dataframe", that is really up to you to code that part.

**Reply by sebastiansukstorf** [reply]
Hello, no but I will have a closer look...Thank you.

**Comment by sebastiansukstorf** [comment]
Hi @mcanouil, I tried the dashboard with this code. But I cant insert the last row C¢ and R$. And I cant mix columns with row due to the orientation in yaml header

```
    ---
    title: "BMC Dashboard"
    author: Sebastian Sukstorf
    format: 
        dashboard:
            orientation: columns
    ---

    ## Column {height=70%}

    ```{python}
    #| title: Chart A
    print("Chart A")
    ```

    ## Column {height=30%}

    ```{python}
    #| title: Chart B
    print("Chart B")
    ```

    ```{python}
    #| title: Chart C
    print("Chart C")
    ```

    ## Column {height=70%}

    ```{python}
    #| title: Chart D
    print("Chart D")
    ```

    ## Column {height=30%}

    ```{python}
    #| title: Chart E
    print("Chart E")
    ```

    ```{python}
    #| title: Chart F
    print("Chart F")
    ```

    ## Column {height=70%}

    ```{python}
    #| title: Chart G
    print("Chart G")
    ```
```

With quarto check:

```
Quarto 1.4.551
[✓] Checking versions of quarto binary dependencies...
      Pandoc version 3.1.11: OK
      Dart Sass version 1.69.5: OK
      Deno version 1.37.2: OK
[✓] Checking versions of quarto dependencies......OK
[✓] Checking Quarto installation......OK
      Version: 1.4.551
      Path: /Applications/quarto/bin

[✓] Checking tools....................OK
      TinyTeX: v2022.12.12
      Chromium: (not installed)

[✓] Checking LaTeX....................OK
      Using: TinyTex
      Path: /Users/sebastiansukstorf/Library/TinyTeX/bin/universal-darwin
      Version: 2022

[✓] Checking basic markdown render....OK

[✓] Checking Python 3 installation....OK
      Version: 3.9.13 (Conda)
      Path: /Users/sebastiansukstorf/opt/anaconda3/bin/python
      Jupyter: 4.11.1
      Kernels: python3

[✓] Checking Jupyter engine render....OK

[✓] Checking R installation...........OK
      Version: 4.3.3
      Path: /Library/Frameworks/R.framework/Resources
      LibPaths:
        - /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library
      knitr: 1.45
      rmarkdown: 2.25

[✓] Checking Knitr engine render......OK
```

**Reply by mcanouil** [reply]
Could you properly format your post using code blocks for code and terminal outputs? Thanks.
If your code contains code blocks, you need to enclose it using more backticks, *i.e.*, usually four ` ```` `.

This being said, you can nest the components, see <https://github.com/mcanouil/quarto-dashboard>.

---

You can share a **self-contained "working" (reproducible)** Quarto document using the following syntax, _i.e._, using more backticks than you have in your document (usually four ` ```` `).
See <https://quarto.org/bug-reports.html#small-is-beautiful-aim-for-a-single-document-with-10-lines>.

If you have multiple files (and if it is **absolutely required** to have multiple files), please share as a Git repository.

<table>
<tr><th>R</th><th>Python</th></tr>
<tr><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: knitr
---

This is a reproducible Quarto document.

```{r}
x <- c(1, 2, 3, 4, 5)
y <- c(1, 4, 9, 16, 25)

plot(x, y)
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td><td>

`````md
````qmd
---
title: "Reproducible Quarto Document"
format: html
engine: jupyter
---

This is a reproducible Quarto document.

```{python}
import matplotlib.pyplot as plt

x = [1, 2, 3, 4, 5]
y = [1, 4, 9, 16, 25]

plt.plot(x, y)
plt.show()
```

![An image](https://placehold.co/600x400.png)

The end.
````
`````

</td></tr>
</table>

Additionally and if not already given, please share the output of `quarto check` within a code blocks (*i.e.*, using three backticks ` ```txt `), see <https://quarto.org/bug-reports.html#check>.

**Reply by sebastiansukstorf** [reply]
Hello @mcanouil, thank you for your remarks. The quarto Preview will generate 
![bmc-quarto-preview](https://github.com/quarto-dev/quarto-cli/assets/6294081/91df4579-78fc-46d2-a3c9-28d6028e95f4)
the attached image.

**Reply by mcanouil** [reply]
I would suggest to go with the row layout and put columns in each row.
Take a look at the repository I linked in my previous reply.

**Reply by sebastiansukstorf** [reply]
![Bildschirmfoto 2024-03-14 um 13 58 24](https://github.com/quarto-dev/quarto-cli/assets/6294081/bd260145-422d-41fa-a6fc-bf000e52ed9a)

Hello @mcanouil I could produce the dashboard with this code:

```
    ---
    title: "BMC Dashboard"
    author: Sebastian Sukstorf
    format: 
        dashboard:
            theme: lux
            embed-resources: true
    engine: knitr
    ---

    ## Row {height="70%"}

    ### Column {width="100%"}

    ::: {.card title="Key Partners" icon="circles-overlap"}
    ```{python}
    print("Gibt es Partnerorganisationen oder Lieferanten, die für den Erfolg Ihres Geschäftsmodells entscheidend sind? Wie 
    können Sie diese Beziehungen weiter stärken?")
    ```
    :::

    ### Column {width="100%"}

    ::: {.card title="Key Activities" icon="circle-check"}
    ```{python}
    print("Welche Aktivitäten sind entscheidend für den Erfolg Ihrer Organisation? Welche Prozesse müssen optimiert werden, 
    um effizienter zu arbeiten?")
    ```
    :::

    ::: {.card title="Key Resources" icon="industry"}
    ```{python}
    print("Welche Ressourcen benötigen Sie, um Ihr Geschäftsmodell erfolgreich umzusetzen? Das können materielle, 
    finanzielle oder auch immaterielle Ressourcen sein.")
    ```
    :::

    ### Column {width="100%"}

    ::: {.card title="Value Propositions" icon="gift"}
    ```{python}
    print("Was bietet Ihre Organisation Ihren Kunden an? Was sind die wichtigsten Produkte oder Dienstleistungen, die Sie 
    bereitstellen? Welche Probleme lösen sie für Ihre Kunden, und welchen Mehrwert bieten sie?")
    ```
    :::

    ### Column {width="100%"}

    ::: {.card title="Customer Relationships" icon="heart"}
    ```{python}
    print("Wie pflegen Sie die Beziehungen zu Ihren Kunden? Welche Art von Interaktion haben Sie mit ihnen, und wie können 
    Sie diese Beziehungen verbessern, um die Kundenzufriedenheit zu steigern?")
    ```
    :::

    ::: {.card title="Channels" icon="truck"}
    ```{python}
    print("Wie gelangen Ihre Produkte oder Dienstleistungen zu Ihren Kunden? Welche Vertriebskanäle nutzen Sie, um Ihre 
    Zielgruppe zu erreichen? Gibt es Möglichkeiten, diese Kanäle zu optimieren oder zu erweitern?")
    ```
    :::

    ### Column {width="100%"}

    ::: {.card title="Customer Segments" icon="face-smile"}
    ```{python}
    print("Wer sind Ihre wichtigsten Kundengruppen und was sind ihre Bedürfnisse und Probleme? Welche unterschiedlichen 
    Segmente können wir identifizieren, und wie können wir sie am besten ansprechen?")
    ```
    :::

    ## Row {height="20%"}

    ::: {.card title="Cost Structure" icon="receipt"}
    ```{python}
    print("Welche sind Ihre wichtigsten Kostenarten? Wie können Sie Kosten senken oder optimieren, um Ihre Rentabilität zu 
    verbessern?")
    ```
    :::

    ::: {.card title="Revenue Streams" icon="circle-euro"}
    ```{python}
    print("Woher generieren Sie Ihr Einkommen? Welche sind Ihre wichtigsten Einnahmequellen, und gibt es Möglichkeiten, 
    diese zu diversifizieren oder zu maximieren?")
    ```
    :::

```

I am facing two problems. I tried to embed icons (is the source https://fontawesome.com/search?) via e.g. {.card title="Revenue Streams" icon="circle-euro"}. 
The other problem is that my embedded text in each .card will not have a line break. Is that possible to achieve?

**Reply by mcanouil** [reply]
Default icons are Bootstrap Icons (BI), see <https://quarto.org/docs/dashboards/data-display.html#icon-and-color>.

Your `print()` calls don't have any linebreaks. Why would you expect Quarto to add some?
Looking at the screenshot, are you actually talking about line wrapping? If so, that's CSS. Use your favourite browser developer/inspect mode to targets the proper elements, and add a new (S)CSS rule to your dashboard.

**Reply by sebastiansukstorf** [reply]
Hello @mcanouil, I tried to embed icons from Bootstrap if I want to work with quarto. In the documentation of quarto dashboards I found that icons can be embedded in a .valuebox. 

```
    ```{python}
    #| content: card
    #| title: "Renevue Streams"
    #| icon: "currency-euro"
    print("Woher generieren Sie Ihr Einkommen? Welche sind Ihre wichtigsten Einnahmequellen, und gibt es Möglichkeiten, 
    diese zu diversifizieren oder zu maximieren?")
    ```
```
But I am using .card elements. I am wondering how I can embed an icon in a .card element? Are the any preconditions for the use of Bootstrap with quarto and what will be the best solution?

**Reply by sebastiansukstorf** [reply]
Hello @mcanouil , For the text wrapping I was given the tip to use a css ore a scss file. I tried with the a css solution `css: custom_style.css`. 

The css file looks like this:
```scss
       div {
            overflow-wrap: break-word;
       }
```

But this solution is not working and I don't know where the error is? Thank you for your help. For your information: I am not a programmer or an IT specialist. I just want to use quarto for my use case if possible.

**Reply by mcanouil** [reply]
Please don't ping me in every single reply. Thank you!

**Reply by mcanouil** [reply]
1. Don't indent the code when you share it, it makes it unusable by others, see <https://github.com/quarto-dev/quarto-cli/discussions/9051#discussioncomment-8776507>
2. If you intend to output text from a code cell, you need to tell Quarto it's not code, it's actual markdown that you output. Use `output: asis` or IPython. Code as no wrapping by default and here your cards contents are not text, but strings.

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/94cfc72a-95ee-469e-a1e0-11b52815e7c0">


Regarding CSS and HTML, I'am afraid, you'll have to learn by yourself using online resources as this is out of Quarto scope.

**Comment by sebastiansukstorf** [comment]
My BMC dashboard working fine. But last I want to know, if I can embed an bootstrap icon in the .card header, just like the the bellow solution. [Bootstrap 4 : Icon in a card header](https://stackoverflow.com/questions/56255210/bootstrap-4-icon-in-a-card-header)

**Reply by mcanouil** [reply]
One way using the [Iconify Quarto extension](https://github.com/mcanouil/quarto-iconify), but you can use raw HTML, etc.
Again, this is really a HTML/CSS question.

You can also write the example you are referring to using fenced divs attributes: https://quarto.org/docs/authoring/markdown-basics.html#divs-and-spans

````qmd
---
title: "BMC Dashboard"
author: Sebastian Sukstorf
format: dashboard
---

::: {.card}
## Value Propositions {{< iconify exploding-head >}}

```{python}
print("Was bietet Ihre Organisation Ihren Kunden an? Was sind die wichtigsten Produkte oder Dienstleistungen, die Sie bereitstellen? Welche Probleme lösen sie für Ihre Kunden, und welchen Mehrwert bieten sie?")
```
:::
````

**Reply by sebastiansukstorf** [reply]
My code is working, but not with all icons from iconify.

```
    ---
    title: "BMC Dashboard"
    author: Sebastian Sukstorf
    format: dashboard
    ---

    ::: {.card}
    ## Key Partners {{< iconify link  >}}

    ```{python}
    #| output: asis
    print("Gibt es Partnerorganisationen oder Lieferanten, die für den Erfolg Ihres Geschäftsmodells entscheidend sind? Wie 
    können Sie diese Beziehungen weiter stärken?")
    ```
    :::


    ::: {.card}
    ## Key Partners {{< iconify fa-solid:link size=lg >}}

    ```{python}
    #| output: asis
    print("Gibt es Partnerorganisationen oder Lieferanten, die für den Erfolg Ihres Geschäftsmodells entscheidend sind? Wie 
    können Sie diese Beziehungen weiter stärken?")
    ```
    :::
```

I tried with ` {{< iconify link  >}}`and it is working. But with other icons from iconify (eg. `{{< iconify fa-solid:link size=lg >}}`there is no output generated.

**Reply by mcanouil** [reply]
The format/name of the icon set / icon name is incorrect, please read the documentation at <https://github.com/mcanouil/quarto-iconify?tab=readme-ov-file#using>.

**Reply by sebastiansukstorf** [reply]
Done! The Business Model Canvas is ready. This is the qmd code:

`````
---
title: "Business Model Canvas"
format: 
    dashboard:
        theme: default
        embed-resources: true
        css: custom_style.css
engine: knitr
---

## Row {height="70%"}

### Column {width="100%"}

::: {.card}
## Key Partners {{< iconify fa-solid link >}}

```{python}
#| output: asis
print("Gibt es Partnerorganisationen oder Lieferanten, die für den Erfolg Ihres Geschäftsmodells entscheidend sind? Wie können Sie diese Beziehungen weiter stärken?")
```
:::

### Column {width="100%"}

::: {.card}
## Key Activities {{< iconify vaadin check-square >}}

```{python}
#| output: asis
print("Welche Aktivitäten sind entscheidend für den Erfolg Ihrer Organisation? Welche Prozesse müssen optimiert werden, um effizienter zu arbeiten?")
```
:::

::: {.card}
## Key Resources {{< iconify wpf factory >}}

```{python}
#| output: asis
print("Welche Ressourcen benötigen Sie, um Ihr Geschäftsmodell erfolgreich umzusetzen? Das können materielle, finanzielle oder auch immaterielle Ressourcen sein.")
```
:::

### Column {width="100%"}

::: {.card}
## Value Propositions {{< iconify heroicons gift-solid >}}

```{python}
#| output: asis
print("Was bietet Ihre Organisation Ihren Kunden an? Was sind die wichtigsten Produkte oder Dienstleistungen, die Sie bereitstellen? Welche Probleme lösen sie für Ihre Kunden, und welchen Mehrwert bieten sie?")
```
:::

### Column {width="100%"}

::: {.card}
## Customer Relationships {{< iconify mdi heart >}}

```{python}
#| output: asis
print("Wie pflegen Sie die Beziehungen zu Ihren Kunden? Welche Art von Interaktion haben Sie mit ihnen, und wie können Sie diese Beziehungen verbessern, um die Kundenzufriedenheit zu steigern?")
```
:::

::: {.card}
## Channels {{< iconify mdi truck >}}

```{python}
#| output: asis
print("Wie gelangen Ihre Produkte oder Dienstleistungen zu Ihren Kunden? Welche Vertriebskanäle nutzen Sie, um Ihre Zielgruppe zu erreichen? Gibt es Möglichkeiten, diese Kanäle zu optimieren oder zu erweitern?")
```
:::

### Column {width="100%"}

::: {.card}
## Customer Segments {{< iconify foundation torsos-all-female >}}

```{python}
#| output: asis
print("Wer sind Ihre wichtigsten Kundengruppen und was sind ihre Bedürfnisse und Probleme? Welche unterschiedlichen Segmente können wir identifizieren, und wie können wir sie am besten ansprechen?")
```
:::

## Row {height="20%"}

::: {.card}
## Cost Structure {{< iconify ant-design tags-filled >}}

```{python}
#| output: asis
print("Welche sind Ihre wichtigsten Kostenarten? Wie können Sie Kosten senken oder optimieren, um Ihre Rentabilität zu verbessern?")
```
:::

::: {.card}
## Revenue Streams {{< iconify healthicons money-bag >}}
```{python}
#| output: asis
print("Woher generieren Sie Ihr Einkommen? Welche sind Ihre wichtigsten Einnahmequellen, und gibt es Möglichkeiten, diese zu diversifizieren oder zu maximieren?")
```
:::

`````

And this is my `custom_style.css` for the background-color:

````
 body {
    background-color: #E7E9EB;
}
````

The rendered html looks like:

![Bildschirmfoto 2024-03-21 um 16 41 08](https://github.com/quarto-dev/quarto-cli/assets/6294081/1a61f2f0-4ee2-44c1-b140-7c944a5d7370)

**Reply by mcanouil** [reply]
@sebastiansukstorf Glad you finally succeeded to get the desired output.
May I ask you to remove the indentation of the code which means any users will have to remove it in order to make it work. Use extra backticks (four, five or more) to enclose your code.
If you don't see the issue with how you share code, I invite you to use the copy button and paste it in a document, then try to render without editing.
![image](https://github.com/quarto-dev/quarto-cli/assets/8896044/88d81b00-afd9-4c9d-a7a9-8ff1cf263b5f)

**Reply by sebastiansukstorf** [reply]
The BMC as a dashboard should now integrated in a qmd document or a Revealjs presentation. How can I integrate the rendered html? I want to integrate the whole bmc (with icons, background-color) and render the new output document as pdf.

**Reply by mcanouil** [reply, marked as answer]
You can't. A dashboard is an output format on its own.
The only place you can really embed a dashboard is a website project.

This being said, you can use technology such as [decktape](https://github.com/astefanutti/decktape), or since you are using R/knitr [webshot2](https://github.com/rstudio/webshot2) to turn your dashboard into a static PDF/PNG which you can then include however you want/like.

(There are way more technologies to turn HTML into PDF/PNG/etc. that you can easily find on GitHub and/or any web search engine)
